paper-server/src/main/java/org/bukkit/craftbukkit/CraftWorld.java

```diff
@@ -2322,6 +2322,22 @@ public void sendGameEvent(Entity sourceEntity, org.bukkit.GameEvent gameEvent, V
     }
     // Paper end
 
+    // Paper start - expose potion mixing API
+    @Override
+    public ItemStack mixPotion(ItemStack potion, ItemStack ingredient) {
+        return CraftItemStack.asCraftMirror(
+            getHandle().potionBrewing().mix(CraftItemStack.asNMSCopy(ingredient), CraftItemStack.asNMSCopy(potion))
```

Comment:
These do not need to be copies, they won't be mutated.
We can save a copy by using `unwrap`.